### PR TITLE
LITE-27887 Add GET /deployments/<id> endpoint

### PR DESCRIPTION
**Pull request by Hairash**
* Refactoring. Make get_deployment_schema function
* Add errors.py

### CODE REVIEW

File: connect_ext_ppr/webapp.py [hunk 1]
```diff
@@ -25,13 +25,34 @@
...
+    )
+    def get_deployment(
+        self,
+        deployment_id: str,
+        client: ConnectClient = Depends(get_installation_client),
+        db: VerboseBaseSession = Depends(get_db),
+    ):
+        dep = db.query(Deployment).filter(Deployment.id == deployment_id).first()
+        if dep is None:
+            raise HTTPException(status_code=404, detail=f'Deployment {deployment_id} not found')
```
Comment (jonatrios) [inline connect_ext_ppr/webapp.py:49]: Here we can use `status` from fastapi and say that `status_code=status.HTTP_404_NOT_FOUND` , it make response/code more clear

File: connect_ext_ppr/webapp.py [hunk 2]
```diff
@@ -25,13 +25,34 @@
...
+        summary='Deployment details',
+        response_model=DeploymentSchema,
+    )
+    def get_deployment(
+        self,
+        deployment_id: str,
+        client: ConnectClient = Depends(get_installation_client),
+        db: VerboseBaseSession = Depends(get_db),
+    ):
+        dep = db.query(Deployment).filter(Deployment.id == deployment_id).first()
```
Comment (jonatrios) [inline connect_ext_ppr/webapp.py:47]: I found that if we do not filter by account, we are going to retrieve a Deployment that should not be visible for the account that is making the request, so I will ask for some refactor:

connect_ext_ppr/models/deployment.py -> Remove `unique=True` constraint from `product_id` field of `Deployment` model.
connect_ext_ppr/models/utils.py -> Add `account_id=installation['owner']['id']` to filter query. Line: 9
connect_ext_ppr/events.py - > Add `account_id=installation['owner']['id']` to filter query. Note: installation in events is available as instance attr `self.installation`. Lines: 43, 82
connect_ext_ppr/webapp.py  Add `installation: dict = Depends(get_installation),` to `get_deployments` route, and add `account_id=installation['owner']['id']` to filter query. Lines: 46
You can also use `filter_by` that allow you to specify kwargs as Django filter: `db.query(Deployment).filter_by(account_id=self.installation['owner']['id']`, product=request['product']['id'])`

**Review by jonatrios — CHANGES_REQUESTED** [review]
About the failure on test, please change type hint `dict`  in [here](https://github.com/cloudblue/connect-extension-xvs/blob/master/connect_ext_ppr/schemas.py#L30) to be `Dict` of `typing`, apparently is not supported for 3.8 version

**Comment by d3rky** [comment]
Guys, if you are making changes in API, please change the original doc also... And please use plurals in the resource's name. I mean
/deployment --> /deploymentS

File: connect_ext_ppr/webapp.py (hunk 1 above)
Comment (Hairash) [reply connect_ext_ppr/webapp.py]: Decided to use ClientError instead

File: connect_ext_ppr/webapp.py (hunk 2 above)
Comment (Hairash) [reply connect_ext_ppr/webapp.py]: Fixed

**Comment by Hairash** [comment]
> Guys, if you are making changes in API, please change the original doc also... And please use plurals in the resource's name. I mean /deployment --> /deploymentS

We use plurals, there was a typo in the commit description, sorry.

File: connect_ext_ppr/webapp.py [hunk 3]
```diff
@@ -37,6 +39,37 @@
 )
 class ConnectExtensionXvsWebApplication(WebApplicationBase):
 
+    @router.get(
```
Comment (rahulmondal) [inline connect_ext_ppr/webapp.py:42]: could you please add some test cases?

**Review by jonatrios — CHANGES_REQUESTED** [review]

File: connect_ext_ppr/webapp.py [hunk 4]
```diff
@@ -37,6 +39,37 @@
...
+    ):
+        dep = (
+            db.query(Deployment)
+            .filter_by(id=deployment_id, account_id=installation['owner']['id'])
+            .first()
+        )
+        if dep is None:
+            raise ObjectNotFound(deployment_id)
+
+        listings = get_all_info(client)
```
Comment (rahulmondal) [inline connect_ext_ppr/webapp.py:62]: May be this is not part of this PR but the name of the function 'get_all_info' seems not descriptive enough.

File: connect_ext_ppr/errors.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+from connect.client import ClientError
+
+
+class ObjectNotFound(ClientError):
+    def __init__(self, obj_id):
+        self.error_code = 'EXT_001'
+        self.status_code = 404
+        self.errors = [f'Object {obj_id} not found']
```
[the remark this answers is not in the document]
Comment (Hairash) [reply connect_ext_ppr/errors.py]: Thanks! Fixed.

File: connect_ext_ppr/webapp.py (hunk 4 above)
Comment (rahulmondal) [inline connect_ext_ppr/webapp.py:62]: It does not make sense to collect all listing, marketplace and hub data for the account and only use a very specific part of it.
We can just get data using Connect API for specific product and hub that is associated with the deployment object.

**Review by d3rky — CHANGES_REQUESTED** [review]

File: connect_ext_ppr/utils.py [hunk 6]
```diff
@@ -55,3 +57,38 @@ def filter_object_list_by_id(object_list, key):
...
+        id=deployment.id,
+        account_id=deployment.account_id,
+        hub={
+            'id': deployment.hub_id,
+            'name': hub['name'],
+        },
+        product={
+            'id': deployment.product_id,
+            'name': product['name'],
+            'icon': product.get('icon', None),
```
Comment (rahulmondal) [inline connect_ext_ppr/utils.py:81]:
```suggestion
            'icon': product.get('icon'),
```

File: connect_ext_ppr/utils.py [hunk 7]
```diff
@@ -55,3 +57,38 @@ def filter_object_list_by_id(object_list, key):
...
+        },
+        owner={
+            'id': deployment.vendor_id,
+            'name': vendor['name'],
+            'icon': vendor.get('icon', None),
```
Comment (rahulmondal) [inline connect_ext_ppr/utils.py:86]:
```suggestion
            'icon': vendor.get('icon'),
```

**Review by rahulmondal — CHANGES_REQUESTED** [review]

File: connect_ext_ppr/webapp.py (hunk 3 above)
Comment (Hairash) [reply connect_ext_ppr/webapp.py:42]: Yes, but we have separate task for it to not block the development

Comment (rahulmondal) [reply connect_ext_ppr/webapp.py:42]: @d3rky 
do you agree with this?
If yes, please specify the ticket for that and resolve the comment.

File: connect_ext_ppr/webapp.py (hunk 4 above)
Comment (jonatrios) [reply connect_ext_ppr/webapp.py:62]: Ok, about [function name](https://github.com/cloudblue/connect-extension-xvs/pull/3#discussion_r1253264370), will change that in another PR. And the information that we collect, I agree with you that in case of list deployments we can change this behaviour, but for first time creation of Deployment objects we need all listing "listed", so we can leave it as it is for now.

[the remark this answers is not in the document]
Comment (jonatrios) [reply connect_ext_ppr/webapp.py:62]: @rahulmondal already comment [here](https://github.com/cloudblue/connect-extension-xvs/pull/3#discussion_r1253283974) retrieving/listing data for the particular `product_id`/ `hub_id` associated with the Deployment object/s that user is trying to retrieve/list, I totally agree with that. But we have in representation of Deployment a lot of fields that could be changed at any moment, like `vendor_name, vendor_icon, product_name, product_icon, distributor_name, distributor_icon, hub_name`, all that fields we have to request. Another thing that we've talked with Rahul was to have some kind of cache, but Im not sure if this is in the scope of the MVP of this extension.

[the remark this answers is not in the document]
Comment (Hairash) [reply connect_ext_ppr/webapp.py:62]: Good question. The only reason I see - is to avoid data duplication. I thought this moment was already discussed. If not - let's return to the discussion.
To make a decision we need to know the use case. How big amount of data are we going to store and how often are we going to change this data? If there will be a lot of deployment but changes will be rare so better not to store the data. In the opposite case - better to store.
But also worth to mention that to catch every change of every field Jonathan wrote above we need more event handling work.

File: connect_ext_ppr/webapp.py (hunk 3 above)
Comment (Hairash) [reply connect_ext_ppr/webapp.py:42]: Decided to add tests there

File: connect_ext_ppr/webapp.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hairash) [reply connect_ext_ppr/webapp.py:62]: Discussed. Decided to leave as is now

Comment (Hairash) [reply connect_ext_ppr/webapp.py:62]: Decided to fix it later

File: connect_ext_ppr/webapp.py (hunk 3 above)
Comment (Hairash) [reply connect_ext_ppr/webapp.py:43]: Added

File: connect_ext_ppr/utils.py (hunk 6 above)
Comment (Hairash) [reply connect_ext_ppr/utils.py]: Fixed

File: connect_ext_ppr/utils.py (hunk 7 above)
Comment (Hairash) [reply connect_ext_ppr/utils.py]: Fixed

File: connect_ext_ppr/webapp.py (hunk 4 above)
Comment (Hairash) [reply connect_ext_ppr/webapp.py]: I changed the way we get information from Connect related to the only deployment.
But now when we use Connect `GET /products/{product_id}` endpoint, we receive the full product's data, and there are a lot of redundant information. Maybe there is a better solution if we need only object's reference?

File: tests/test_events.py [hunk 8]
```diff
@@ -26,11 +26,13 @@ def test_handle_listing_processing(
 ):
     config = {}
     listing['status'] = status
-    client_mocker = client_mocker_factory(base_url=connect_client.endpoint)
 
-    client_mocker.marketplaces.filter(R().id.in_([marketplace['id']])).limit(1).mock(
-        return_value=[marketplace],
-    )
+    if status == 'listed':
```
Comment (jonatrios) [inline tests/test_events.py:30]: [COD-C] Here instead of an `if` statement we must have separate tests for the ones who makes calls with connect client for listed / installed and the others who don't like `test_unlisted_do_nothing` / `test_uninstalled_do_nothing` , and remove the marker of parametrize. But we can leave this as an improvement, wdyt?

**Review by jonatrios — APPROVED** [review]

**Review by rahulmondal — APPROVED** [review]
